# Week 1 Activity: More with Numpy, Indexing, and Playing Sound

Complete this activity as part of your participation grade. Pending length of the lecture, you will have time in class to work. Everything you need to complete this activity can be found in this week's (or a previous week's) lecture code.

### Part 1: Loading, playing, and plotting an audio file. 

1) Read in a .wav file (from the server) and assign it a variable name. (You will need to import a Python library)

In [5]:
from scipy.io.wavfile import read
fs, x = read("../audio/Flute-A4.wav")



2) Playback your audio file

In [26]:
?read

Signature: read(filename, mmap=False)
Docstring:
Open a WAV file.

Return the sample rate (in samples/sec) and data from an LPCM WAV file.

Parameters
----------
filename : string or open file handle
    Input WAV file.
mmap : bool, optional
    Whether to read data as memory-mapped (default: False).  Not compatible
    with some bit depths; see Notes.  Only to be used on real files.

    .. versionadded:: 0.12.0

Returns
-------
rate : int
    Sample rate of WAV file.
data : numpy array
    Data read from WAV file. Data-type is determined from the file;
    see Notes.  Data is 1-D for 1-channel WAV, or 2-D of shape
    (Nsamples, Nchannels) otherwise. If a file-like input without a
    C-like file descriptor (e.g., :class:`python:io.BytesIO`) is
    passed, this will not be writeable.

Notes
-----
Common data types: [1]_

=====================  ===========  ===========  =============
     WAV format            Min          Max       NumPy dtype
=====================  ===========  ====

3) Figure out the length of the .wav file in samples and its duration in seconds. If your file is longer than 5 seconds, truncate the file so it is less than 5 seconds in length.

In [11]:
num_samples = x.size
duration_seconds = num_samples / fs

num_samples, duration_seconds




(94803, 2.149727891156463)


94803 samples, 94803/44100 = 2.15 seconds duration

4) Make a plot of your file (or segment of your file). **Make sure the x-axis shows time in seconds**, label your axes, and provide a title.

### Part 2: Manipulating the audio through NumPy functions, indexing, and boolean masking.

1) Reverse all the values in your audio array. Save in a new variable, "reverse", and play it back. 

In [13]:
reverse = x[::-1]
from IPython.display import Audio
Audio(reverse, rate=fs)

2) Randomize all the sample values in your original audio file and save in a new variable, "random", and use each sample exactly once. Play back the first 1 second.

In [ ]:

random = np.random.permutation(x)

Audio(random[:fs], rate=fs)


3) Remove every other sample from the original file, and save this as a new variable, "skipped". Play it back and compare to the original. Try different sample rates for playback.

*Hint: there are two ways to do this, one way removes values from the original and the other indexes from the original*

In [18]:
skipped = x[::2]
Audio(skipped, rate=fs)
Audio(skipped, rate=fs // 2)   # closer to original pitch
Audio(skipped, rate=fs * 2)    # chipmunk chaos




4) Repeat the above, except only keep every fifth sample from the original. Play it back and compare.

In [ ]:
every_fifth = x[::5]
Audio(every_fifth, rate=fs)


5) What have you effectively done by removing values?

In [ ]:
removing values downsamples the signal, reducing resolution, causes information loss and aliasing

6) Using your original audio file, find the position of everywhere in the file where the amplitude is greater than some value (you get to decide the value.) 

In [20]:
threshold = 10000
mask = np.abs(x) > threshold
positions = np.where(mask)
len(positions[0])




0

7) Reassign each sample that matches your 'greater than x' criteria to a new value, then playback the new array, "masked". (Be careful when setting values!)

In [ ]:
masked = x.copy()
masked[mask] = 0
Audio(masked, rate=fs)




8) Using your original audio's sampling rate, figure out how many samples you would need to define a "grain" of sound that is 100ms long. Make one grain from the beginning of the original audio file, "start".

In [23]:
grain_samples = int(0.1 * fs)  
start = x[:grain_samples]     

Audio(start, rate=fs)


9) Using the audio arrays you have created above (reversed, masked, skipped, etc), create more grains of 50-100ms lengths then concatenate and repeat (i.e. continuously append) until you have two seconds worth of sound. Play your new sound back.

In [25]:
sources = [x, reverse, masked, skipped]
target_len = int(2 * fs)
out = []
current = 0
while current < target_len:
    src = sources[np.random.randint(len(sources))]

    grain_len = np.random.randint(int(0.05 * fs), int(0.10 * fs))

    if src.size <= grain_len:
        start_i = 0
    else:
        start_i = np.random.randint(0, src.size - grain_len)

    grain = src[start_i:start_i + grain_len]
    out.append(grain)
    current += grain.size

new_sound = np.concatenate(out)[:target_len]
Audio(new_sound, rate=fs)